echo.go

```diff
@@ -517,8 +517,8 @@ func (e *Echo) File(path, file string, m ...MiddlewareFunc) *Route {
 func (e *Echo) add(host, method, path string, handler HandlerFunc, middleware ...MiddlewareFunc) *Route {
 	name := handlerName(handler)
 	router := e.findRouter(host)
+	h := applyMiddleware(handler, append(e.middleware, middleware...)...)
```

Comment:
Besides of logic changes in this PR, it seems that moving the call to applyMiddleware() outside the handler that is added to the router, could be a small performance improvement, because that will be performed only one time when the route is added and not in every call to the route when Echo is started. This improvement is more important, for Groups, because for every route in a Group, this method is called with all the middlewares of the Group plus the ones for that particular route.

I created a quick&dirty benchmark in master branch just moving the applyMiddleware() as it is in master and the gain in performance is small but the benchmark shows also that this change reduces the number of allocations

Change that I measured...
```go
	h := applyMiddleware(handler, middleware...)
	router.Add(method, path, func(c Context) error {
		return h(c)
	})
```

...benchmark code...
```go
func dummyMiddleware(next HandlerFunc) HandlerFunc {
	return func(c Context) error {
		return next(c)
	}
}

func BenchmarkGroups(b *testing.B) {
	e := New()
	g := e.Group("/group")

	g.Use(dummyMiddleware, dummyMiddleware, dummyMiddleware, dummyMiddleware, dummyMiddleware, dummyMiddleware)
	g.GET("/foo", func(c Context) error {
		return c.String(http.StatusOK, c.Path())
	}, dummyMiddleware, dummyMiddleware, dummyMiddleware, dummyMiddleware, dummyMiddleware, dummyMiddleware)

	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		request(http.MethodGet, "/group/foo", e)
	}
}
```

...benchmarks compared with benchstat
```
name      old time/op    new time/op    delta
Groups-2    4.00µs ± 2%    3.60µs ± 5%  -10.00%  (p=0.000 n=9+10)

name      old alloc/op   new alloc/op   delta
Groups-2    5.95kB ± 0%    5.76kB ± 0%   -3.23%  (p=0.000 n=10+10)

name      old allocs/op  new allocs/op  delta
Groups-2      31.0 ± 0%      19.0 ± 0%  -38.71%  (p=0.000 n=10+10)
```